Forall Type over Coercion

xnning opened the pull request:
Goal:

- Add forall-type-over-coercion

xnning committed:
var

xnning committed:
data con

xnning committed:
mkid

xnning committed:
rename TyVarBndr -> VarBndr, TvBndr -> Bndr in Var

xnning committed:
TyCoRep (WIP)

xnning committed:
went through TyCoRep for the first time

goldfirere commented on compiler/types/TyCoRep.hs:
```diff
@@ -737,23 +738,33 @@ infixr 3 `mkFunTy`      -- Associates to the right
+  -- we don't need to do this because we consider co -> ty to be equivalent as
+  -- \/_ : co. ty
+
+mkFreshCoVarOfType :: InScopeSet -> Type -> CoVar
```
I'm hoping we never need this in the end.

goldfirere commented on compiler/types/TyCoRep.hs:
```diff
@@ -2526,6 +2522,35 @@ substForAllCoBndrUsing sym sco (TCvSubst in_scope tenv cenv)
+    new_kind_co | no_kind_change = old_kind_co
+                | otherwise      = sco old_kind_co
+
+    -- TODO: unsure
```
So am I. (Unsure, that is) It would be helpful to illustrate this with an example, connecting the pieces of the example back to the variables here.

goldfirere reviewed (commented):
I've reviewed Var.hs and TyCoRep.hs. They look good!

xnning committed:
TyCon

xnning committed:
add comments according to the discussion

xnning committed:
revert the change: tyConTyCoVar -> tyConTyVars in TyCon

xnning committed:
typo

xnning committed:
remove extra space introduced before revert

xnning committed:
finish DataCon

xnning committed:
zipTCvSubst

xnning committed:
Type.hs

xnning committed:
the note Weird typing rule for ForAllTy

xnning committed:
mkforallty/mkpity

xnning committed:
clean up ToCoRep

xnning committed:
MkId

xnning committed:
hEqTyCon -> heqTyCon

xnning committed:
CoAxiom

xnning committed:
Revert "CoAxiom"

This reverts commit 0964fe7ff0a152b4b6b2595eaf0219d2317798f0.

xnning committed:
coercion.hs

xnning committed:
instenv.hs

xnning committed:
change ForAllCo rule according to meeting; and other changes.

xnning committed:
Revert "instenv.hs"

This reverts commit eef8c0d90e48710c33eebefbd27c79e1d3ddf6a6.

xnning committed:
ifacetype

xnning committed:
ifacesyn

xnning committed:
ToIface

xnning committed:
Fixes in Type, TyCoRep

xnning committed:
Fix TyCoRep and Type

xnning committed:
PatSyn

xnning committed:
Fixes in Coercion

xnning committed:
unify.hs

xnning committed:
Fixes in DataCon, Type, Coercion

xnning committed:
TysWiredIn

xnning committed:
add comment for ty_co_match for coercion variables

xnning committed:
Fix Coercion according to FamInstEnv

xnning committed:
FamInstEnv

xnning committed:
Fix IfaceSyn

xnning committed:
CoreUtils

xnning committed:
PatSyn

xnning committed:
ConLike

xnning committed:
typo in CoreUtils

xnning committed:
TcType

xnning committed:
CoreSubst

xnning committed:
small fix in Type

xnning committed:
OptCoercion

xnning committed:
clean up OptCoercion

xnning committed:
adjust Coercion for OptCoercion; add missing cases (for NthCo and InstCo) in Coercion

xnning committed:
CoreArity

xnning committed:
add test functions to distinguish ForAllTy over ty/co vars in Type

xnning committed:
Small changes for CoreMap, TysPrim, CoreSubst, IfaceSyn

xnning committed:
add exports to make MkId work

xnning committed:
correct ToIface

xnning committed:
small fix in ToIface

xnning committed:
BuildTyCl

xnning committed:
RnModIface

xnning committed:
CoreOpt

xnning committed:
small refactor in OptCoercion

xnning committed:
only renaming changes in TcMType

xnning committed:
CoreTidy

xnning committed:
Renaming in TcRnTypes

xnning committed:
MkIface

xnning committed:
CoreLint

xnning committed:
SimplEnv

xnning committed:
TcIface

xnning committed:
small changes in TyCoRep and Type

xnning committed:
make all renaming stuff under deSugar/ ghci/ typecheck/ to make it compilable and don't care about what's happening

xnning committed:
correct Coercion; add annotations in OptCoercion

xnning committed:
fix TyCoRep

xnning committed:
small fix

xnning committed:
fix IfaceType.hs-boot

xnning committed:
remove isCoVar in Coercion and redundant import of Var in SpecConstr

xnning committed:
remove unused function in TcIFace

xnning committed:
revert changes in PatSyn: pattern will not have exvars.

xnning committed:
add fix

xnning committed:
fix TcDerivUtils

xnning committed:
cosmetic changes

xnning committed:
revert PatSyn: ex only has tyvars

xnning committed:
CoreLint

xnning committed:
export substTyVarBndr in Type

xnning committed:
clean note for VarBndrs, TyCoVarBinders, TyConBinders...

xnning committed:
CoreUtils

xnning committed:
cosmetic changes; typos; support substTyVarBndr in SimplEnv

xnning committed:
substTyVarBndrs is available in TcDerivInfer now

xnning committed:
under typechecking

xnning committed:
Coercion

xnning committed:
simplify OptCoercion

xnning committed:
TyCoRep

xnning committed:
Small fix in CoreLint

xnning committed:
mkPiCo_unchecked under typecheck

xnning committed:
add unchecked functions

xnning committed:
add Assert for mkPiCo

xnning committed:
Fix comments in DataCon: correct annotations; add an example for coercion variables

xnning committed:
address more comments: typos and renaming

xnning committed:
Note for unsed coercion variable in ForAllTy

xnning committed:
annotation for VarBndr, typos, and Assert in lintCoercion

xnning committed:
add note for mkForAllCo

xnning committed:
Add comments for instCoercion and notes for liftCoSubstVarBndr

xnning committed:
correct DsExpr

xnning committed:
update haddock

xnning committed:
Address comments in DataCon

xnning committed:
add ASSERT for smart constructors for ForAllCo; correct typos/annotations

xnning committed:
relocate ASSERTs

xnning committed:
Fix parsing errors

xnning committed:
address some comments: add ASSERTs, add TyVarBinder

xnning committed:
userTyCoBinders -> userTyBinders

xnning committed:
revert dependent GADT equalities in dataConFullSig

xnning committed:
More comments; and some simplification of cases

xnning committed:
add comments for mkPiCo; revert explicitly printing forall for covars

xnning committed:
remove mkForAllCo_uncheck, use mkForAllTy for the kind of ForAllCo

xnning committed:
more changes on revert explicitly printing foralls

xnning committed:
the inner_kind of ForAllCo over covar is TYPE r

xnning committed:
refactor liftCoSubstVarBndrUsing

xnning committed:
rename uncheck related functions

xnning committed:
add TyConTyCoBinder

xnning committed:
changes regarding names

xnning committed:
final clean up; add Note extendLiftingContextEx; TcSplitxxx only produce tyvars

xnning committed:
Fix comments: remote typo and add some comments

mgsloan committed:
Add support for ImplicitParams and RecursiveDo in TH

Summary:
This adds TH support for the ImplicitParams and RecursiveDo extensions.

I'm submitting this as one review because I cannot cleanly make
the two commits independent.

Initially, my goal was just to add ImplicitParams support, and
I found that reasonably straightforward, so figured I might
as well use my newfound knowledge to address some other TH omissions.

Test Plan: Validate

Reviewers: goldfire, austin, bgamari, RyanGlScott

Reviewed By: RyanGlScott

Subscribers: carter, RyanGlScott, thomie

GHC Trac Issues: #1262

Differential Revision: https://phabricator.haskell.org/D1979

alpmestan committed:
tests: increase (compile) timeout multiplier for T13701 and MultiLayerModules

Summary:
Those tests are currently making our i386 validation fail on CircleCI:

  https://circleci.com/gh/ghc/ghc/8827

Test Plan: Using my Phab<->CircleCI bridge to run i386 validation for this diff.

Reviewers: bgamari, monoidal

Reviewed By: monoidal

Subscribers: rwbarton, carter

GHC Trac Issues: #15484, #15383

Differential Revision: https://phabricator.haskell.org/D5103

monoidal committed:
Fix T15502 on 32-bit

Summary:
The expected output uses a hardcoded value for
maxBound :: Int.

This should fix one of circleci failures on i386.

Test Plan: make test TEST=T15502

Reviewers: RyanGlScott, bgamari

Reviewed By: RyanGlScott

Subscribers: rwbarton, carter

GHC Trac Issues: #15502

Differential Revision: https://phabricator.haskell.org/D5151

last-g committed:
Mark system and internal symbols as private symbols in asm

Summary:
This marks system and internal symbols as private in asm output so those
random generated sysmbols won't appear in .symtab

Reasoning:
 * internal symbols don't help to debug because names are just random
 * the symbols style breaks perf logic
 * internal symbols can take ~75% of the .symtab. In the same time
   .symtab can take about 20% of the binary file size

Notice:
This diff mostly makes sense on top of the D4713 (or similar)

Test Plan:
tests

Perf from D4713
```
     7.97%  FibbSlow  FibbSlow            [.] c3rM_info
     6.75%  FibbSlow  FibbSlow            [.] 0x000000000032cfa8
     6.63%  FibbSlow  FibbSlow            [.] cifA_info
     4.98%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_eqIntegerzh_info
     4.55%  FibbSlow  FibbSlow            [.] chXn_info
     4.52%  FibbSlow  FibbSlow            [.] c3rH_info
     4.45%  FibbSlow  FibbSlow            [.] chZB_info
     4.04%  FibbSlow  FibbSlow            [.] Main_fibbzuslow_info
     4.03%  FibbSlow  FibbSlow            [.] stg_ap_0_fast
     3.76%  FibbSlow  FibbSlow            [.] chXA_info
     3.67%  FibbSlow  FibbSlow            [.] cifu_info
     3.25%  FibbSlow  FibbSlow            [.] ci4r_info
     2.64%  FibbSlow  FibbSlow            [.] s3rf_info
     2.42%  FibbSlow  FibbSlow            [.] s3rg_info
     2.39%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_eqInteger_info
     2.25%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_minusInteger_info
     2.17%  FibbSlow  FibbSlow            [.] ghczmprim_GHCziClasses_zeze_info
     2.09%  FibbSlow  FibbSlow            [.] cicc_info
     2.03%  FibbSlow  FibbSlow            [.] 0x0000000000331e15
     2.02%  FibbSlow  FibbSlow            [.] s3ri_info
     1.91%  FibbSlow  FibbSlow            [.] 0x0000000000331bb8
     1.89%  FibbSlow  FibbSlow            [.] ci4N_info
...
```

Perf from this patch:
```
    15.37%  FibbSlow  FibbSlow            [.] Main_fibbzuslow_info
    15.33%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_minusInteger_info
    13.34%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_eqIntegerzh_info
     9.24%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_plusInteger_info
     9.08%  FibbSlow  FibbSlow            [.] frame_dummy
     8.25%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_eqInteger_info
     4.29%  FibbSlow  FibbSlow            [.] 0x0000000000321ab0
     3.84%  FibbSlow  FibbSlow            [.] stg_ap_0_fast
     3.07%  FibbSlow  FibbSlow            [.] ghczmprim_GHCziClasses_zeze_info
     2.39%  FibbSlow  FibbSlow            [.] 0x0000000000321ab7
     1.90%  FibbSlow  FibbSlow            [.] 0x00000000003266b8
     1.88%  FibbSlow  FibbSlow            [.] base_GHCziNum_zm_info
     1.83%  FibbSlow  FibbSlow            [.] 0x0000000000326915
     1.34%  FibbSlow  FibbSlow            [.] 0x00000000003248cc
     1.07%  FibbSlow  FibbSlow            [.] base_GHCziNum_zp_info
     0.98%  FibbSlow  FibbSlow            [.] 0x00000000003247c8
     0.80%  FibbSlow  FibbSlow            [.] 0x0000000000121498
     0.79%  FibbSlow  FibbSlow            [.] stg_gc_noregs
     0.75%  FibbSlow  FibbSlow            [.] 0x0000000000321ad6
     0.67%  FibbSlow  FibbSlow            [.] 0x0000000000321aca
     0.64%  FibbSlow  FibbSlow            [.] 0x0000000000321b4a
     0.61%  FibbSlow  FibbSlow            [.] 0x00000000002ff633
```

Reviewers: simonmar, niteria, bgamari

Reviewed By: simonmar

Subscribers: lelf, angerman, olsner, rwbarton, thomie, carter

Differential Revision: https://phabricator.haskell.org/D4722

last-g committed:
Mark code related symbols as @function not @object

Summary:
This diff is a part of the bigger project which goal is to improve
common profiling tools support (perf) for GHC binaries.

A similar job was already done and reverted in the past:
 * https://phabricator.haskell.org/rGHCb1f453e16f0ce11a2ab18cc4c350bdcbd36299a6
 * https://phabricator.haskell.org/rGHCf1f3c4f50650110ad0f700d6566a44c515b0548f

Reasoning:

`Perf` and similar tools build in memory symbol table from the .symtab
section of the ELF file to display human-readable function names instead
of the addresses in the output. `Perf` uses only two types of symbols:
`@function` and `@notype` but GHC is not capable to produce any
`@function` symbols so the `perf` output is pretty useless (All the
haskell symbols that you can see in `perf` now are `@notype` internal
symbols extracted by mistake/hack).

The changes:
 * mark code related symbols as @function
 * small hack to mark InfoTable symbols as code if TABLES_NEXT_TO_CODE is true

Limitations:
 * The perf symbolization support is not complete after this patch but
   I'm working on the second patch.
 * Constructor symbols are not supported. To fix that we can issue extra
   local symbols which mark code sections as code and will be only used
   for debug.

Test Plan:
tests
any additional ideas?

Perf output on stock ghc 8.4.1:
```
     9.78%  FibbSlow  FibbSlow            [.] ckY_info
     9.59%  FibbSlow  FibbSlow            [.] cjqd_info
     7.17%  FibbSlow  FibbSlow            [.] c3sg_info
     6.62%  FibbSlow  FibbSlow            [.] c1X_info
     5.32%  FibbSlow  FibbSlow            [.] cjsX_info
     4.18%  FibbSlow  FibbSlow            [.] s3rN_info
     3.82%  FibbSlow  FibbSlow            [.] c2m_info
     3.68%  FibbSlow  FibbSlow            [.] cjlJ_info
     3.26%  FibbSlow  FibbSlow            [.] c3sb_info
     3.19%  FibbSlow  FibbSlow            [.] cjPQ_info
     3.05%  FibbSlow  FibbSlow            [.] cjQd_info
     2.97%  FibbSlow  FibbSlow            [.] cjAB_info
     2.78%  FibbSlow  FibbSlow            [.] cjzP_info
     2.40%  FibbSlow  FibbSlow            [.] cjOS_info
     2.38%  FibbSlow  FibbSlow            [.] s3rK_info
     2.27%  FibbSlow  FibbSlow            [.] cjq0_info
     2.18%  FibbSlow  FibbSlow            [.] cKQ_info
     2.13%  FibbSlow  FibbSlow            [.] cjSl_info
     1.99%  FibbSlow  FibbSlow            [.] s3rL_info
     1.98%  FibbSlow  FibbSlow            [.] c2cC_info
     1.80%  FibbSlow  FibbSlow            [.] s3rO_info
     1.37%  FibbSlow  FibbSlow            [.] c2f2_info
...
```

Perf output on patched ghc:
```
     7.97%  FibbSlow  FibbSlow            [.] c3rM_info
     6.75%  FibbSlow  FibbSlow            [.] 0x000000000032cfa8
     6.63%  FibbSlow  FibbSlow            [.] cifA_info
     4.98%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_eqIntegerzh_info
     4.55%  FibbSlow  FibbSlow            [.] chXn_info
     4.52%  FibbSlow  FibbSlow            [.] c3rH_info
     4.45%  FibbSlow  FibbSlow            [.] chZB_info
     4.04%  FibbSlow  FibbSlow            [.] Main_fibbzuslow_info
     4.03%  FibbSlow  FibbSlow            [.] stg_ap_0_fast
     3.76%  FibbSlow  FibbSlow            [.] chXA_info
     3.67%  FibbSlow  FibbSlow            [.] cifu_info
     3.25%  FibbSlow  FibbSlow            [.] ci4r_info
     2.64%  FibbSlow  FibbSlow            [.] s3rf_info
     2.42%  FibbSlow  FibbSlow            [.] s3rg_info
     2.39%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_eqInteger_info
     2.25%  FibbSlow  FibbSlow            [.] integerzmgmp_GHCziIntegerziType_minusInteger_info
     2.17%  FibbSlow  FibbSlow            [.] ghczmprim_GHCziClasses_zeze_info
     2.09%  FibbSlow  FibbSlow            [.] cicc_info
     2.03%  FibbSlow  FibbSlow            [.] 0x0000000000331e15
     2.02%  FibbSlow  FibbSlow            [.] s3ri_info
     1.91%  FibbSlow  FibbSlow            [.] 0x0000000000331bb8
     1.89%  FibbSlow  FibbSlow            [.] ci4N_info
...
```

Reviewers: simonmar, niteria, bgamari, goldfire

Reviewed By: simonmar, bgamari

Subscribers: lelf, rwbarton, thomie, carter

GHC Trac Issues: #15501

Differential Revision: https://phabricator.haskell.org/D4713

xnning committed:
update haddock
